crates/runix/src/store_path.rs

```diff
@@ -0,0 +1,60 @@
+use std::fmt::Display;
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+
+use once_cell::sync::Lazy;
+use thiserror::Error;
+
+/// Respect [NIX_STORE_DIR](https://nixos.org/manual/nix/stable/command-ref/env-common.html#env-NIX_STORE_DIR)
+static STORE_PREFIX: Lazy<String> = Lazy::new(|| {
+    std::env::var("NIX_STORE_DIR")
+        .ok()
+        .or_else(|| option_env!("NIX_STORE_DIR").map(String::from))
+        .unwrap_or_else(|| "/nix/store".to_string())
+});
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct StorePath(PathBuf);
+
+impl TryFrom<PathBuf> for StorePath {
+    type Error = StorePathError;
+
+    fn try_from(value: PathBuf) -> Result<Self, Self::Error> {
+        let path = value
+            .as_path()
+            .strip_prefix(STORE_PREFIX.as_str())
+            .map_err(|_| StorePathError::NotAStorePath(value.clone()))?;
+
+        path.components()
+            .next()
```

Comment:
So does this drop extra path components? That's a bit confusing. If you prefer that behavior and don't want to error, I think it definitely merits a comment on the function, and possibly even a different function name than `try_from` that makes it more obvious it's lossy

Reply:
yeah right now it strips those
but youre right we might want to keep more of the path around and have accessors for the other parts